mac80211: minstrel: fix variance overestimation / fluctuation

The variance / standard deviation is artificially inflated for rates
where only sample attempts are counted. With only few packet attempts,
the difference between the current probabilty and the ewma value tends
to be very big because of the lack of precision in the current value.

Whenever not enough data was gathered, pass in the previous average
instead of the current probability. This leads to the standard deviation
value slowly decreasing for sampled rates.

Signed-off-by: Felix Fietkau <nbd@openwrt.org>
---
 net/mac80211/rc80211_minstrel.c | 11 ++++++++---
 net/mac80211/rc80211_minstrel.h |  2 ++
 2 files changed, 10 insertions(+), 3 deletions(-)

--- a/net/mac80211/rc80211_minstrel.c
+++ b/net/mac80211/rc80211_minstrel.c
@@ -159,7 +159,7 @@ minstrel_update_rates(struct minstrel_pr
 void
 minstrel_calc_rate_stats(struct minstrel_rate_stats *mrs)
 {
-	unsigned int cur_prob;
+	unsigned int cur_prob, ewmv_prob;
 
 	if (unlikely(mrs->attempts > 0)) {
 		mrs->sample_skipped = 0;
@@ -167,11 +167,16 @@ minstrel_calc_rate_stats(struct minstrel
 		if (unlikely(!mrs->att_hist)) {
 			mrs->prob_ewma = cur_prob;
 		} else {
+			ewmv_prob = cur_prob;
+
+			if (mrs->attempts < 12)
+				ewmv_prob = mrs->prob_ewma;
+
 			/* update exponential weighted moving variance */
 			mrs->prob_ewmv = minstrel_ewmv(mrs->prob_ewmv,
-							cur_prob,
+							ewmv_prob,
 							mrs->prob_ewma,
-							EWMA_LEVEL);
+							EWMV_LEVEL);
 
 			/*update exponential weighted moving avarage */
 			mrs->prob_ewma = minstrel_ewma(mrs->prob_ewma,
--- a/net/mac80211/rc80211_minstrel.h
+++ b/net/mac80211/rc80211_minstrel.h
@@ -10,7 +10,9 @@
 #define __RC_MINSTREL_H
 
 #define EWMA_LEVEL	96	/* ewma weighting factor [/EWMA_DIV] */
+#define EWMV_LEVEL	121
 #define EWMA_DIV	128
+
 #define SAMPLE_COLUMNS	10	/* number of columns in sample table */
 
 /* scaled fraction values */
